codegen/src/test/resources/software/amazon/awssdk/codegen/poet/auth/scheme/query-endpoint-auth-params-with-allowlist-auth-scheme-interceptor.java

```diff
@@ -0,0 +1,130 @@
+        executionAttributes.putAttribute(SdkInternalExecutionAttribute.SELECTED_AUTH_SCHEME, selectedAuthScheme);
+    }
+
+    private List<AuthSchemeOption> resolveAuthOptions(Context.BeforeExecution context,
+      ExecutionAttributes executionAttributes) {
+        QueryAuthSchemeProvider authSchemeProvider = Validate.isInstanceOf(QueryAuthSchemeProvider.class, executionAttributes.getAttribute(SdkInternalExecutionAttribute.AUTH_SCHEME_RESOLVER), "Expected an instance of QueryAuthSchemeProvider");
+        QueryAuthSchemeParams params = authSchemeParams(context.request(), executionAttributes);
+        return authSchemeProvider.resolveAuthScheme(params);
+  }
+
+    private SelectedAuthScheme<? extends Identity> selectAuthScheme(List<AuthSchemeOption> authOptions,
+            ExecutionAttributes executionAttributes) {
+        Map<String, AuthScheme<?>> authSchemes = executionAttributes.getAttribute(SdkInternalExecutionAttribute.AUTH_SCHEMES);
+        IdentityProviderConfiguration identityResolvers = executionAttributes
+                .getAttribute(SdkInternalExecutionAttribute.IDENTITY_PROVIDER_CONFIGURATION);
+        List<Supplier<String>> discardedReasons = new ArrayList<>();
+        for (AuthSchemeOption authOption : authOptions) {
+            if (authOption.schemeId().equals("smithy.api#noAuth")) {
+                if (!discardedReasons.isEmpty()) {
+                    LOG.debug(() -> String.format("%s auth will be used, discarded: '%s'", authOption.schemeId(),
+                            discardedReasons.stream().map(Supplier::get).collect(Collectors.joining(", "))));
+                }
+                return new SelectedAuthScheme<>(null, null, authOption);
+            }
+            AuthScheme<?> authScheme = authSchemes.get(authOption.schemeId());
+            SelectedAuthScheme<? extends Identity> selectedAuthScheme = trySelectAuthScheme(authOption, authScheme,
+                    identityResolvers, discardedReasons);
+            if (selectedAuthScheme != null) {
+                if (!discardedReasons.isEmpty()) {
+                    LOG.debug(() -> String.format("%s auth will be used, discarded: '%s'", authOption.schemeId(),
+                            discardedReasons.stream().map(Supplier::get).collect(Collectors.joining(", "))));
+                }
+                return selectedAuthScheme;
+            }
+        }
+        throw SdkException
+                .builder()
+                .message(
+                        "Failed to determine how to authenticate the user: "
+                                + discardedReasons.stream().map(Supplier::get).collect(Collectors.joining(", "))).build();
+    }
+
+    private QueryAuthSchemeParams authSchemeParams(SdkRequest request, ExecutionAttributes executionAttributes) {
+        QueryEndpointParams endpointParams = QueryResolveEndpointInterceptor.ruleParams(request, executionAttributes);
```

Comment:
Just thinking aloud (to confirm my understanding, not actionable for now), so with this, it would look like, for S3:
1. S3AuthSchemeInterceptor creates S3EndpointParams (via calling S3ResolveEndpointInterceptor.ruleParams)
2. converts S3EndpointParams to S3AuthSchemeParams
3. DefaultS3AuthSchemeProvider.resolveAuthScheme converts S3AuthSchemeParams back to S3EndpointParams
4. S3ResolveEndpointInterceptor creates another S3EndpointParams instance (2nd time)

Reply:
Yes, I was going to replicate the logic to generate the params here but the amount of code generated is non-trivial, so this way felt much cleaner, even though all this params translations have to happen.

Reply:
yeah, duplicating the code is one option, it avoids one conversion, but still creates S3EndpointParams twice. another option I considered is extracting the `ruleParams()` into a S3CreateEndpointParamsInterceptor that is run first and saves the S3EndpointParams for use by both S3AuthSchemeInterceptor and S3ResolveEndpointInterceptor - doesn't avoid any additional conversion, but avoids creating another instance. I was thinking through it, because there is desire to avoid running resolveEndpoint twice, and this is in the slightly related. But probably not worth doing any optimization from what you have. This is not as expensive and just putting data into appropriate objects. So I think what you have is fine.

Reply:
There are a couple of things about this that aren't optimal - calling it twice as @gosar points out, but also that we're calling a public method on another interceptor, which feels gross. For the first one, I don't know either that we need to optimize it now. We don't have data on how long it takes to resolve the params or resolve the endpoint yet. For the second one, I'm thinking it'd be nice to extract the rule param creation to it's own component, but I'm not super familiar with the other conversions that are happening that Jay mentioned so I don't have an explicit suggestion for now. It can also wait.

Reply:
`ruleParams` seems pretty inexpensive, at least for query in this one code review, so I'm not too upset about calling it once for auth and once for endpoints.

If we were hand-writing the code, I think I'd favor putting the creation of the params into a different class. But since this is code generation it means a lot more to maintain in the code generator just so that the generated code is cleaner/more intuitive. I'd favor keeping the code we have to maintain simpler than the code we generate.